test/alex_koin/slack_commands_test.exs

```diff
@@ -0,0 +1,122 @@
+defmodule AlexKoin.SlackCommandsTest do
+  use AlexKoin.DataCase
+
+  alias AlexKoin.SlackCommands
+  alias AlexKoin.Repo
+  alias AlexKoin.Account.{Transaction, User, Wallet}
+  alias AlexKoin.Coins.Coin
+
+  @memo "All of the reasons"
+
+  describe "create_coin" do
+    setup do
+      user = insert(:user)
+
+      %{
+        user: user,
+        wallet: insert(:wallet, user: user),
+        reason: Faker.Lorem.sentence()
+      }
+    end
+
+    test "creates a coin with a reason", %{user: user, reason: reason} do
+      %User{id: user_id} = user
+
+      assert {:ok, %Coin{} = new_coin} = SlackCommands.create_coin(user, user, reason)
+      assert new_coin.origin == reason
+      assert new_coin.created_by_user_id == user_id
+      assert new_coin.mined_by_id == user_id
+    end
+
+    test "fails with invalid reason", %{user: user} do
+      assert {:error, :new_coin, _, _} = SlackCommands.create_coin(user, user, nil)
+    end
+  end
+
+  describe "remove_coins" do
+    setup do
+      %{
+        wallet: insert(:wallet)
+      }
+    end
+
+    test "errors when there are no coins", %{wallet: wallet} do
+      assert {:error, :not_enough_coins} = SlackCommands.remove_coins(wallet, 1)
+    end
+
+    test "errors when there are not enough coins", %{wallet: wallet} do
+      insert(:coin, wallet: wallet)
+      assert {:error, :not_enough_coins} = SlackCommands.remove_coins(wallet, 2)
+    end
+
+    test "success if there are enough coins", %{wallet: wallet} do
+      insert(:coin, wallet: wallet)
+      assert :ok = SlackCommands.remove_coins(wallet, 1)
+    end
+  end
+
+  describe "transfer" do
+    setup context do
+      coin_count = Map.get(context, :coin_count, 1)
+      source_wallet = insert(:wallet, balance: coin_count)
+
+      %{
+        source_wallet: source_wallet,
+        target_wallet: insert(:wallet),
+        coins: insert_list(coin_count, :coin, wallet: source_wallet)
+      }
+    end
+
+    def assert_balances([]), do: true
+    def assert_balances([{wallet, amount} | rest]) do
+      float_amount = amount / 1
+
+      assert %Wallet{balance: ^float_amount} = Repo.get(Wallet, wallet.id)
+      assert amount == wallet |> Coin.for_wallet() |> Repo.all |> length
+
+      assert_balances(rest)
+    end
+
+    @tag coin_count: 0
```

Comment:
Question: what does this do?

Reply:
It's kind of like exunit's version of let statements.  It will merge this into the context for this test. So you can pass in things per test and write a common setup block to handle it.